### PR TITLE
property textures in Custom Shaders

### PR DESCRIPTION
**Pull request by ptrgags**
:construction:  WIP :construction: 

This PR adds property textures from the `EXT_mesh_features` to Custom Shaders. This first PR is for basic support and adds every property texture to the shader, but in a future PR, I'll check which textures are actually used in the custom shader (similar to what I did in CustomShaderPipelineStage)

To Do:
- [ ] Add another test dataset or two for unit tests. Consider the model from https://github.com/CesiumGS/3d-tiles-samples/pull/40, but maybe add another with a `vec2` property
- [ ] Consider adding a Sandcastle demo
- [ ] Add unit tests for `MetadataPipelineStage

[Local Sandcastle for testing](http://localhost:8080/Apps/Sandcastle/index.html#c=7TuLcts4kr+C9dbVUIlMSX7kIcepdfzIZMueuGInubo45YJISEIMElyAlKXZyr9fNwCSoES/4tzNztZOZWQJBLob/e4GuM80L5LwcJ4xxROW5lQcMZoXiumQpXQk2ImMmfCfk12Sq4LtXKQzqsiMs2umYCxl12TfQvtkxoKLtcj83pewjKdMXax1yT8vUkKkipl6l8YsY/CR5ueKploUEUujxZCMqdCse5F+7wAOwGKghZGQ0VUYFUrhAqAFcDp8fy8Ep+kBzVk4VjJ5p+WLZ/1BgJgu1jb6G4P1wWC9//J8oz/ceD7cfB72B8/6z7a2nm9vbr149mLr5ebz/7lYu0gtvl6PmD2TTGqec5nydEJ2H/afoRu4U4Koad2nKodvNN00tB6wiWJMG1rXBxubYf/51tazwcsuDmxthf3t/ubz/jPzs28pRLjTbInlvzIaA52nPI+mH6QQQb9L8F+5YMznLD5SNGGG12OpkpqmakiHwGTQADNPHlVr3jKQHs2lcjxNpcqnIE3765rpvMG+/ULnMiFnUwpiJgdszFPDBU0ewj8AdKDoZILsz6eMJLLQjFxzIQibZzSNe3qqeHrlHoLEQrtX+9TI0JHQYJWlzj4JjDoWKTe7H1rtJARQf2IR7JegjIgAzdI5iYEaAjqaA8Rc1jTZNcUlPq9AEJIvMjYskX60KM5hLPx0uL/RLWfNqChgmk9fqSEbQT9EGYb9jpv+3fy1nzMG0+Z2G+dsng/Jl4r4d2OPY1wbyicsLqlWfDLNuzXfyETJa33f1YKNG4utFNzyi7WZ5LEj7gSsPvhkvr5LsyInM23+dglPJfyMfk8uDQw7532R20n2S8cpGAL9p/fdcM3NCUsDO9knT3eRVQPyxIkinMNXhzGkea74qADGhqC8CRUn+zsezO/u+9fwmwSaQZtBA9c6nhMCvQIS0TE6Uf64ajFBM438RCfGU6JZJNNYE5hIaMoTihsqlQon3Uepjo7f750va1W/oTcGOchUrcdokkiB3VGJSwOPMqbvg+5s7+T0+PDD5cbBLZp83mBYUEEFXEoMge1h2IN/ZzTJBAP3TXs2YJSEhFk6qYQEu2gzA9jSXpaJhVFIt5/aOkG3ugTEDl9nrDElkhCEeIqW7cBQMuI5gXnKSkZLwvNftBMJi0OcNwbFwjDYYndO9csZRvmP3A+r/uMb1P8EwCsOkTVxX25VfRZt4C72cQOgfuNVDS+fXvbJU7AJNAlcFVhtsh5lZwlsiTqM+XiMhr9bsmrjIKhUo1th7oRqMnqQCQGLE3rFgM/RlEVXTI0kVZUOgmPPp8BrQkU2pSAqmgObIsWoZto+BPEZKDGnE5kCu9IiGUFWkY6LNDJRFuHve8CdAgYOxxn/nXWscqM9N8iwdvyRp/mLPaXoItgCrnnrmr/sjiyYhM4PSop2yQZM9PGRdTLA2cSYd4ALOEzr78CfVz5IGHj6tFOaXjX5m538bXnyN2+ypYNDKjWH2XyJ8Kfk2045D4Mz7rqMqRlVPF8QaT3+EmPJhM+A9015eZAyUDmm0tAnIq45wRuY+ZgE1cP/AjZBlIf0hHhOAUBqlhtK9D8KCjqhwOrIZ0ZkCgaeMhuDyjnwECijacrEX2ogPq1fUIaGLV9RMtvbFTXfvW1IAKauOeYW7Jd4mQRAOBIUMk/yBkyWomoAS+ABFaCb8cIDZLIcKoDnhs6+RheSAnjkNlIP44CKhReptwoljfis3rsNdUGrslaZAOAvkBTAZr6WQNq3DNzfxG0HMBXGSjgd0vM11nHE8MN8KAaKk97txTEyxMZShg0CnD+45nE+Hfp66B5MGSYfbU8w35PpL2j2oFaZxKyrCzIh1zQFZ6m4zpomy+IJc2lHArwf88hEziMuYP1wifyTlRnhb4d7Hw7PyqCZ0MmdMFanNIAYR/fdJgvRqhsCYdzkoF5ULvKjtnGqsVNP1QlFi0yovlrFcwKjP5iWFJc+oMdlAC1bbwncf7JIurO6eGs5gvgBs2ETddS8R9hN+DxYHu4ivs3AlAINrKHqPDQI70sBLkfw1VBsEiUM0SZRwqIQ4gPW5OjaFyYIGwi+X1rVwj10ZH8aNYTtpAzSTm12OGWYjuc5euxKBsYz41yu60YFujBVsP+o8m2q3NTUZr7YWGKD39IC+lDNbvWdEDAj4OtUCshloNqqCoNVzf0V5/zHgf6ptA7zyqaakdemwCHL1OJ/MdcRTFoG8v2BinY+5ZqwualbsUMCOhazRKZQJQHbrA7KdAbcNl2vAqt9zPOk4jgGVW3pRt2OQ/J07gBVDRoohmLydIHDpgv15MlI5thXw0kye/IkrGmxS0cGreY6Byy2ZvrMRm+PfWLMotVNUKEbe9BgMNcIEw2ZEjRpho1XkvE5JCGoxDTvGjjGpgBURDFOfXj7JlyqyN4qGuO+y2THK8E0FiiYmj/bKYfqxHK1JtuERFbbYky7Kqy9rtL3Kqj0PSop3VJCYTpW16bOp4AAY66Y3XbgC7LjuwK7tS+bXprexzS9geCtYhCIVlEsfBRGKRoqcReiASLiPqIR+BtUnr7tduBXmfC87HbUhcGPlgb192ZhoFcrAm/IKNmR0bHKdZ7WYyEo2UrCPWlqmUu2V3SvmQX5HSEqJER9nmvy8ZPfHgqb8H84PpQAHhcblrb5LxkX7ugllVvo3h4wfqjDdNOJAWyCj22T8PTNB8tMbQVrH52O1L/xYcEtOoG2IHh0ZcKNoRrINfn+MrPup0Gzn6FBYyGp5eEBRDOaRqhDgqWTfBpY3q0kAnYJEM6pO3HTiZT5FIJhZnmz2Td/fKgeFOBDvZjGMTZ+yUSKGDwxhOy8lKLOWFQIPIMDtyXQdelW3a/mLVV05XhV0r1AqW3jxwCKu4qGm0hDr6atPpUcnQipNcdOaAsdShaTaQoFDqbXAqJ9MAj7ZL2GaJRm0CWDRle4iRN8L3azEuZin+NFad8ROtNbPcDT3XK7ZqdOU7ExXSJ58HFMBo/yfSGL+DOdsT/fccwdx3j/5ydpZQpo1v2iydx4gIUf+kyj0yUfUPlOmOk9Dr52oVLm0bSCY/NL9OkCUhRZjLDWoi2BdKl2KGYgsJYjuvpUL5wvGjp5UrYnjPRBoUH8INW4QLeLpFLAiBntnFxTc+TjnILJoiowkDZNwZIp+BXcNEo1BG7AL1jL8yI2WdGIol5Iu3/I3SGwUDzvpBmrILnWuU9OYJhTNogREEViFkQXagwoOxYV0leBwfwZ6IbdY7cPNsNn1TGWiyJy7B9qTRk6RXMk4kDk5Sk+tVWHiUZWXmT9NfmyPqi+msFV51lvHlJycBJP7pDN73i4EfbbPbFh/64HE3PpNLBwUV8zTK+LWTgHKHbUmlfHzXS/du4++gVC8Ox325x2YbMbcd+iNppkhdBWXUD6aECjhWmwlqkDt6XJDXgBxJktXAbgxQDrtiH/BqJvdmZ/bK7WwpaIKxuzOJSBoLyoc/YcpCU4rwTmibg8NT7ZtAPWURat5Xq18rWZ8mOFekk9RbYXGlllohDJqGB57iwfj32r8mlibrGIuoZaNmJUG3tcjMbgFb0Qs6uKyNQObDzmEfo7Ta4ZniZPpYY8gaoRBzNUXCzafd343r7uxky6H+IRYwsgs/3L/q19ttVwvIGReBM/tjqgyW2jrYZrJm5iIN/Aj4GN5s6cN31zfnhgVzKDMLZwJc79IvsfX/U4s2CpxiPVWtYJy2lMcxrO2AS+oqwP7KR7SNuw2cH0EqcH83SKtceDGGpSW3CH5h5VleMc+6Phx9+O353/ixSdGH/BbXETvVZ4b5+csyTDi2zAhJ2W5YD8pvXu0S0ArPgBD6YitgBpo8I9va/oDdTAUt/pbWxvowbYQUdSNVoDf/CR0AFLpH7oHccbrjwayWGxYbTI6B1e9ZnmeTbs9cz9QnCW+XDr5fOXvYSqSK5r04PUvcP/Pr9MmJ5ejt0V1N4ZxyenTYfQPhpORD52e014pGQkdYKYe2dQcemeuV9k1Fb33sLUY4ma2jspp/Ym4vyoV630wY2oEFKmrTeW7HZ71jze2InNXwBp5ADFSqbsbjAHOM3/7oHI6HWq7wviFCd7ixMurs5VEV0NS+24HcpJOX/5twdzAgVmGn9ikPMKdifct/7s5i9/k43S7k6gp43pSz8rsN8rZ4h3MTR2itBG8bKxuyQbWzOwjqtsZhpXdrFWahtx6lZZGKTaTEA+MSRV0zvw+sn2odG7wJIbVjrWbQ91napD20bocgG5ROdJIXKOHf1lgvUtFN9BsDHibksMuQelNxLqLgk/np0NBey23tJ8FEeBkRGLC8wcH0+scyXdG+5tPIpQ73C6iif6Z9Pqne4/ilg8qy1b9VQTZMHPJtU7zv1ZClAeJ/xMTWgeLDzO+G2Xu27akhmn5MS0wRHSz7OypaZ5O9XuNZE7iD401/XdOxc/jd4q1HVX3wd4DLV77jYwMVGGmDDzCDKbka7b2tR8iEp8NUGuwu5jLJQA+/Dy/PIc1r5fo8FvsDBTPOHYfNKhYlg97wkRdKrDWXvjf/fGNTSOA0uSqxdW3h4yb71g/lVfB8dMa4hXwqt+qU/lsPGrmlIWCIrPhy3vsUyX3ofx32YJ6tuiZfHdrYemmfJ+OciHQvAMmwzh57dnL7a8CS0v1pQPq4MYcy5bXhm2b6qY1iA4loRrFuZTlga1xpgZldqUrz8BBkXDsVicyzdoirC7swyvk9r54agx2K21Dj0XAh6aY4aSpB17fhpWt18DvDoAHjNkSknlDlbPwHAiqnMYBtGeSylGVJ2wtAhMrlQVEId4vYBMYbZg6gcriTI70zlV5QteoDemo5lGLEzltdHEpu5BKfExi/HVLyDQkHFs7j9AGbtig+blINvtR/hnrtG/S4IVPGS9pgOvzA76/b5hWVt6EWqWl2fgF2u27YHvuq3ismxvM/KHwfAqezzO2otMv3gpm4UH+7Z1vHvDkWHjBTsd4a2DM2yKGz7+aqWJhJkyds+mir6ezhgateMutvb+UnmoSn/trYEd7x6B6fFwCNOxe9VwxaHAwwp61SCzvDNY3E7cayxBA1jnNsw+Fm+/+8a48D1FBb4CZthXH822tcdWd2GwvCKOTPXOeW2F50mjiigre8G3GVMW1CIyQu2WIjpbEoU5Ujo+PDq/PHj/+bda9rAFCg7ORYSbZfz/LWHT5a040fEO2yaOdfZovno98vYjD0u3c4dOs1vfqNTFKFc0yoM6SjjGszQ+Xfb1NferIY+hdqj02oZ462iqQ0Q7vpJdLFtybBMZ4k6tUxe9/PTp7iXf71SPk/cfzw4vT95/OvyDrPoHbcsZyYrxlDZ3H6v4eGr3vNZde6XzhWCvS4H+jSeZVDlmF0EY9nKWZOaCRm8EqSHLw0jrOr36a24DnP9SyohGVzb3HRI1GdFga6NLyv/74Yt6NfZmYozAQ7KVzb3hkXndeR3T/EI3H35fQc1NR9QjAA+AeUTFOhV8AkE84TEIcBXrei6zIdloYC4f2VtizacO9auez7BXMZ8RHu+2vLqNlwe0hifjQgg8/rpYe/2qB/NXlgppcq/3QLmgC5w2Hbw+toNhGL7qwc/2lY4JS5D/Fw) -- note that the second example uses data from https://github.com/CesiumGS/3d-tiles-samples/pull/40

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ptrgags!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/custom-shaders-property-textures/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

File: Source/Scene/MetadataUnpackingStep.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * A namespace that contains several functions for producing GLSL code for
+ * unpacking values on the GPU. These functions take a GLSL expression as a
+ * string and wrap it with more GLSL code to create a new expression, much like
+ * a GLSL macro, but done in JS.
+ *
+ * @namespace MetadataUnpackingStep
+ *
+ * @private
+ */
+const MetadataUnpackingStep = {};
```
Comment (ptrgags) [inline Source/Scene/MetadataUnpackingStep.js:13]: The intent of these unpacking steps is to be composable. Right now I tried to make it so it generates one expression like so:

```
cast("int")(unnormalizeU8("x")) // => int(255.0 * (x))
```

The downside is if you have many steps, the generated line of code could look messy.

Another approach is to generate steps as new lines, something like this: (inspired by https://nodematerial-editor.babylonjs.com/)

```
// original
vec2 temp0 = texture2D(...).rg;
// after unnormalizeU8()
vec2 temp1 = 255.0 * temp0;
// after cast("int")
ivec2 temp2 = int(temp1);
// result
metadata.property = temp2;
```

This approach might allow for more complex initialization logic for things like arrays (see my other comment), but to do this correctly, I need to keep track of the types at every line. Doable, but it feels like unnecessary complexity. 

@lilleyse any thoughts here?

File: Source/Scene/PropertyTextureProperty.js [hunk 2]
```diff
@@ -23,27 +27,29 @@ import GltfLoaderUtil from "./GltfLoaderUtil.js";
...
   //>>includeStart('debug', pragmas.debug);
-  Check.typeOf.object("options.property", property);
+  Check.typeOf.object("options.textureInfo", textureInfo);
+  Check.typeOf.object("options.channels", channels);
   Check.typeOf.object("options.classProperty", classProperty);
-  Check.typeOf.object("options.textures", textures);
+  Check.typeOf.object("options.texture", texture);
   //>>includeEnd('debug');
 
-  const textureInfo = property.texture;
```
Comment (ptrgags) [inline Source/Scene/PropertyTextureProperty.js:36]: Revert some of these changes, but keep the GLSL type / unpacking step changes

File: Specs/Scene/PropertyTextureSpec.js [hunk 3]
```diff
@@ -69,37 +69,20 @@ describe("Scene/PropertyTexture", function () {
...
 
-    const featureTextureJson = {
+    const propertyTextureJson = {
       class: "map",
       extras: extras,
       extensions: extensions,
+      index: 1,
+      texCoord: 1,
       properties: {
-        color: {
```
Comment (ptrgags) [inline Specs/Scene/PropertyTextureSpec.js:77]: These specs probably need to be reverted

File: Specs/Scene/PropertyTexturePropertySpec.js [hunk 4]
```diff
@@ -49,33 +49,26 @@ describe("Scene/PropertyTextureProperty", function () {
       description: "Extra",
     };
 
-    extensions = {
```
Comment (ptrgags) [inline Specs/Scene/PropertyTexturePropertySpec.js:52]: Revert everything except the new functions for unpacking types/steps

**Review by ptrgags — COMMENTED** [review]
After a discussion today about future changes to the schema for property textures, I'm going to have to rework some of this PR

**Comment by ptrgags** [comment]
There's going to be some upcoming `EXT_mesh_features` changes in the next couple weeks. that will affect property textures and other metadata details. I think it's best to wait on this custom shader PR until those are implemented. Until then, I'm closing this PR.

**Pull request closed without merging**